### PR TITLE
[JENKINS-33296] - plugin dependency issues during install

### PR DESCRIPTION
**Pull request by kzantow**
This addresses a few issues with dependent plugins:
- previously "implied" plugin dependencies are not available without explicit dependency references
- maven-plugin was being explicitly excluded as a dependency, causing many failures
- mailer had explicit dependencies in the main codebase for functionality, including saving the 'email' address field for Jenkins user accounts
- dependencies were included in client-side search text in the installer

This addresses:
https://issues.jenkins-ci.org/browse/JENKINS-33296
https://issues.jenkins-ci.org/browse/JENKINS-33308

@jenkinsci/code-reviewers

### CODE REVIEW

File: core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java [hunk 1]
```diff
@@ -379,19 +379,31 @@ private User createAccount(StaplerRequest req, StaplerResponse rsp, boolean self
...
+            }
         }
         user.save();
         return user;
     }
+    
+    public boolean isMailerPluginPresent() {
+        try {
+            // mail support has moved to a separate plugin
+            return null != Jenkins.getInstance().pluginManager.uberClassLoader.loadClass("hudson.tasks.Mailer$UserProperty");
```
Comment (daniel-beck) [inline core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java]: Wouldn't it be easier to query the plugin manager for `mailer`? But probably safer this way…

Comment (slide) [inline core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java]: I wish I had never moved this UserProperty out of core...

**Comment by daniel-beck** [comment]
For reference, while this fixes Maven Plugin related issues, it does not take care of [JENKINS-33308](https://issues.jenkins-ci.org/browse/JENKINS-33308) type dependency problems.

**Comment by kzantow** [comment]
@daniel-beck right, it's only a part of the issue

File: core/src/main/java/hudson/PluginManager.java [hunk 2]
```diff
@@ -1140,7 +1140,9 @@ public HttpResponse doInstallPlugins(StaplerRequest req) throws IOException {
                     String pluginName = n.substring(0, index);
                     String siteName = n.substring(index + 1);
                     UpdateSite.Plugin plugin = getPlugin(pluginName, siteName);
-                    // TODO: Someone that understands what the following logic is about, please add a comment.
+                    // There could be cases like:
+                    // 'plugin.ambiguous.updatesite' where both
+                    // 'plugin' @ 'ambigiuous.updatesite' and 'plugin.ambiguous' @ 'updatesite' resolve to valid plugins
```
Comment (kzantow) [inline core/src/main/java/hudson/PluginManager.java]: Unrelated, but no need for a TODO here

**Comment by kzantow** [comment]
@reviewbybees

File: war/src/main/js/api/plugins.js [hunk 3]
```diff
@@ -19,7 +19,7 @@ exports.recommendedPlugins = [
     "gradle",
     "ldap",
     "mailer",
-    // "matrix-auth",
+    "matrix-auth",
```
Comment (jglick) [inline war/src/main/js/api/plugins.js]: Effectively reverting 7f87486c73b1c63cc595c0fc380d42ada45f6116 I suppose.

File: core/src/main/java/hudson/ClassicPluginStrategy.java [hunk 4]
```diff
@@ -259,6 +259,19 @@ private static Manifest loadLinkedManifest(File archive) throws IOException {
         return new PluginWrapper(pluginManager, archive, manifest, baseResourceURL,
                 createClassLoader(paths, dependencyLoader, atts), disableFile, dependencies, optionalDependencies);
     }
+    
+    /**
+     * Returns all the bundled plugin dependencies for a particular Jenkins version
+     */
+    public static List<PluginWrapper.Dependency> getPreviouslyBundledDependencies(String jenkinsVersion) {
```
Comment (jglick) [inline core/src/main/java/hudson/ClassicPluginStrategy.java]: :ant: the naming is misleading. Whether these were _bundled_ in 1.x or not is irrelevant—only whether they were _detached_ from core. (There are some bundled plugins which were never detached, and we could have unbundled detached plugins though as a matter of policy I think we never did.)

File: core/src/main/java/hudson/ClassicPluginStrategy.java [hunk 5]
```diff
@@ -259,6 +259,19 @@ private static Manifest loadLinkedManifest(File archive) throws IOException {
...
+        List<PluginWrapper.Dependency> out = new ArrayList<>();
+        for (DetachedPlugin detached : DETACHED_LIST) {
+            if (jenkinsVersion == null || jenkinsVersion.equals("null") || new VersionNumber(jenkinsVersion).compareTo(detached.splitWhen) <= 0) {
+                out.add(new PluginWrapper.Dependency(detached.shortName + ':' + detached.requireVersion));
```
Comment (jglick) [inline core/src/main/java/hudson/ClassicPluginStrategy.java]: :bug: reuse `DetachedPlugin.fix`, not only to avoid duplication but so you can honor `BREAK_CYCLES`.

Comment (jglick) [inline core/src/main/java/hudson/ClassicPluginStrategy.java]: (probably after refactoring that method to accept `String yourName, String jenkinsVersion` rather than `Attributes atts`)

File: core/src/main/java/hudson/PluginManager.java [hunk 6]
```diff
@@ -1150,9 +1152,23 @@ public HttpResponse doInstallPlugins(StaplerRequest req) throws IOException {
                     index = n.indexOf('.', index + 1);
                 }
             }
+            
             if (p == null) {
                 throw new Failure("No such plugin: " + n);
             }
+            
+            // JENKINS-33308 - automatically install implied/previously bundled dependencies for older plugins that may need them
+            for(PluginWrapper.Dependency previouslyBundledDependency : ClassicPluginStrategy.getPreviouslyBundledDependencies(p.requiredCore)) {
```
Comment (jglick) [inline core/src/main/java/hudson/PluginManager.java]: This smells like the wrong place to be doing this. The detached plugin mapping should be logically applied to incoming metadata, whether from UC JSON or from a plugin manifest. You seem to be adding in these dependencies at the last minute, whereas they should been considered dependencies from the start.

I suspect that the difference is tangible: if you go into the setup wizard and check the box next to some ancient plugin, the box for the detached dependency like `matrix-project` should get checked then too (which of course _also_ means that we will wind up installing it, but without any special handling _here_).

File: core/src/main/java/hudson/model/UpdateSite.java [hunk 7]
```diff
@@ -634,10 +634,8 @@ public Plugin(String sourceId, JSONObject o) {
             this.categories = o.has("labels") ? (String[])o.getJSONArray("labels").toArray(new String[0]) : null;
             for(Object jo : o.getJSONArray("dependencies")) {
                 JSONObject depObj = (JSONObject) jo;
-                // Make sure there's a name attribute, that that name isn't maven-plugin - we ignore that one -
```
Comment (jglick) [inline core/src/main/java/hudson/model/UpdateSite.java]: WTF

File: core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java [hunk 8]
```diff
@@ -379,19 +379,31 @@ private User createAccount(StaplerRequest req, StaplerResponse rsp, boolean self
...
-        try {
-            // legacy hack. mail support has moved out to a separate plugin
-            Class<?> up = Jenkins.getInstance().pluginManager.uberClassLoader.loadClass("hudson.tasks.Mailer$UserProperty");
-            Constructor<?> c = up.getDeclaredConstructor(String.class);
-            user.addProperty((UserProperty)c.newInstance(si.email));
-        } catch (RuntimeException e) {
-            throw e;
-        } catch (Exception e) {
-            LOGGER.log(Level.WARNING, "Failed to set the e-mail address",e);
+        if(isMailerPluginPresent()) {
```
Comment (jglick) [inline core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java]: Simpler to just catch, and ignore, `ClassNotFoundException` right here.

File: core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java [hunk 9]
```diff
@@ -379,19 +379,31 @@ private User createAccount(StaplerRequest req, StaplerResponse rsp, boolean self
...
+                throw e;
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
         }
         user.save();
         return user;
     }
+    
+    public boolean isMailerPluginPresent() {
```
Comment (jglick) [inline core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java]: :bug: should not be an API:

```java
@Restricted(NoExternalUse.class) // just here for Jelly
```

File: core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java (hunk 1 above)
Comment (jglick) [inline core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java]: :ant: `loadClass` cannot ever return null, so you need not and should not inspect the return value.

```java
try {
    Jenkins.getInstance().pluginManager.uberClassLoader.loadClass("hudson.tasks.Mailer$UserProperty");
    return true;
} catch (ClassNotFoundException e) {
    return false;
}
```

Comment (jglick) [inline core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java]: > I wish I had never moved this UserProperty out of core...

Creating files like `Mailer$UserProperty.java` is technically legal, but pretty messy! Anyway might cause linkage errors now (mismatch in outer class code attribute; cannot remember if this applies only to inner classes or also to nested classes).

File: war/src/main/js/pluginSetupWizardGui.js [hunk 10]
```diff
@@ -599,7 +599,13 @@ var createPluginSetupWizard = function(appendTarget) {
...
 			}
 			else {
-				var matches = findElementsWithText($pl[0], text.toLowerCase(), function(d) { return d.toLowerCase(); });
+				var matches = [];
+				$containers.find('.title,.description').each(function() {
+					var localMatches = findElementsWithText(this, text.toLowerCase(), function(d) { return d.toLowerCase(); });
+					if(localMatches.length > 0) {
+						matches = matches.concat(localMatches);
+					}
+				});
```
Comment (jglick) [inline war/src/main/js/pluginSetupWizardGui.js]: Is this related?

File: core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java [hunk 11]
```diff
@@ -358,7 +358,7 @@ private User createAccount(StaplerRequest req, StaplerResponse rsp, boolean self
         if(si.fullname==null || si.fullname.length()==0)
             si.fullname = si.username;
 
-        if(si.email==null || !si.email.contains("@"))
+        if(isMailerPluginPresent() && (si.email==null || !si.email.contains("@")))
```
Comment (jglick) [inline core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java]: This is really crying out for an extension point—a way for `UserPropertyDescriptor` to indicate that it wishes to be configured during account creation, for security realms supporting such a concept.

File: core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java (hunk 1 above)
Comment (kzantow) [inline core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java]: @jglick right, most of the changes here were just from refactoring

File: war/src/main/js/pluginSetupWizardGui.js (hunk 10 above)
Comment (kzantow) [inline war/src/main/js/pluginSetupWizardGui.js]: Related to issues with dependencies, insomuch as they were erroneously being included in search results in the installer

File: core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java (hunk 11 above)
Comment (kzantow) [inline core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java]: Agree, but given the timing and existing code, I think handling of this is probably sufficient: this at least hides the fields that would not save properly, which would be very confusing and buggy to the user

File: war/src/main/js/api/plugins.js (hunk 3 above)
Comment (kzantow) [inline war/src/main/js/api/plugins.js]: Yes, since the issues are 'fixed'

File: core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java (hunk 11 above)
Comment (jglick) [inline core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java]: Yes I think this is good enough for now, just mentioning it for the record.

Comment (daniel-beck) [inline core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java]: > This is really crying out for an extension point

Agreed, would be great, but out of scope for this PR and 2.0 I think. 2.1 maybe…

File: core/src/main/java/hudson/model/UpdateSite.java [hunk 12]
```diff
@@ -612,7 +623,7 @@ public Api getApi() {
         public final String[] categories;
 
         /**
-         * Dependencies of this plugin.
+         * Dependencies of this plugin, a name -&gt; version mapping.
```
Comment (jglick) [inline core/src/main/java/hudson/model/UpdateSite.java]: BTW `→` requires no escape. :-)

File: core/src/main/java/hudson/ClassicPluginStrategy.java [hunk 13]
```diff
@@ -260,6 +260,40 @@ private static Manifest loadLinkedManifest(File archive) throws IOException {
                 createClassLoader(paths, dependencyLoader, atts), disableFile, dependencies, optionalDependencies);
     }
 
+    private static void fix(Attributes atts, List<PluginWrapper.Dependency> optionalDependencies) {
+        // don't fix the dependency for yourself, or else we'll have a cycle
```
Comment (jglick) [inline core/src/main/java/hudson/ClassicPluginStrategy.java]: This comment might make more sense inside `getImpliedDependencies` where the `pluginName` is used.

File: core/src/main/java/hudson/ClassicPluginStrategy.java [hunk 14]
```diff
@@ -260,6 +260,40 @@ private static Manifest loadLinkedManifest(File archive) throws IOException {
...
+        String jenkinsVersion = atts.getValue("Jenkins-Version");
+        if (jenkinsVersion==null)
+            jenkinsVersion = atts.getValue("Hudson-Version");
+        
+        optionalDependencies.addAll(getImpliedDependencies(pluginName, jenkinsVersion));
+    }
+    
+    /**
+     * Returns all the plugin dependencies that are implicit based on a particular Jenkins version
+     */
```
Comment (jglick) [inline core/src/main/java/hudson/ClassicPluginStrategy.java]: `@since TODO`

Comment (jglick) [inline core/src/main/java/hudson/ClassicPluginStrategy.java]: (assuming you expect this to be public)

Comment (jglick) [inline core/src/main/java/hudson/ClassicPluginStrategy.java]: (BTW for `plugin-compat-tester` we would actually need the list to be present in a static resource. CC @kwhetstone.)

File: core/src/main/java/hudson/model/UpdateSite.java [hunk 15]
```diff
@@ -467,7 +468,17 @@ public boolean isLegacyDefault() {
...
             }
             for(Map.Entry<String,JSONObject> e : (Set<Map.Entry<String,JSONObject>>)o.getJSONObject("plugins").entrySet()) {
-                plugins.put(e.getKey(),new Plugin(sourceId, e.getValue()));
+                Plugin p = new Plugin(sourceId, e.getValue());
+                // JENKINS-33308 - include implied dependencies for older plugins that may need them
+                List<PluginWrapper.Dependency> implicitDeps = ClassicPluginStrategy.getImpliedDependencies(p.name, p.requiredCore);
+                if(!implicitDeps.isEmpty()) {
+                    for(PluginWrapper.Dependency dep : implicitDeps) {
+                        if(!p.dependencies.containsKey(dep.shortName)) {
+                            p.dependencies.put(dep.shortName, dep.version);
```
Comment (jglick) [inline core/src/main/java/hudson/model/UpdateSite.java]: FWIW in `createPluginWrapper` these are added as optional dependencies—I am not really sure why, and that is perhaps part of the problem in JENKINS-33308.

File: core/src/main/java/hudson/ClassicPluginStrategy.java (hunk 14 above)
Comment (kzantow) [inline core/src/main/java/hudson/ClassicPluginStrategy.java]: Is it fine to leave it literally "@since TODO" should I guess at the upcoming version? Sorry, just don't know what to do here, specifically.

**Comment by jglick** [comment]
:bee: but please check behavior w.r.t. disabling implied dependencies which I suspect is still broken. (For example with the Flexible Publish → Matrix Project dep.)

File: core/src/main/java/hudson/ClassicPluginStrategy.java (hunk 14 above)
Comment (jglick) [inline core/src/main/java/hudson/ClassicPluginStrategy.java]: As a rule in PRs we just leave it as `TODO`, with the expectation that the person merging the PR should replace that with the actual version then in the POM. Of course in this case you expect the version to be 2.0 (or some uninteresting alpha variant thereof).

File: core/src/main/java/hudson/model/UpdateSite.java (hunk 15 above)
Comment (kzantow) [inline core/src/main/java/hudson/model/UpdateSite.java]: Yes, they're added as optional dependencies, but this is too late in the process, I believe this just makes the classes available at runtime?

**Comment by kzantow** [comment]
I did test the update center attempting to remove an implied dependency, it treats them as normal, required dependencies and doesn't let you. @jglick

**Comment by jglick** [comment]
> disabling implied dependencies which I suspect is still broken

Yes…so you have fixed the *UI* portion of the code to consider these, and did so as mandatory dependencies. I suspect that the mistake in `PluginManager` (i.e., the non-UI part) treating them as optional will not become visible until we have something like #2001. You would expect for example that

    touch $JENKINS_HOME/work/plugins/matrix-project.jpi.disabled

and restarting would cause Flexible Publish to be disabled, too. Even if there were such a defense in core—today there is not—the inexplicable use of `optionalDependencies` for detached dependencies means that it would still not work for these cases.

Thus not a regression in your PR, but rather an apparent bug in code your PR refactored.

File: core/src/main/java/hudson/ClassicPluginStrategy.java [hunk 16]
```diff
@@ -260,6 +260,41 @@ private static Manifest loadLinkedManifest(File archive) throws IOException {
                 createClassLoader(paths, dependencyLoader, atts), disableFile, dependencies, optionalDependencies);
     }
 
+    private static void fix(Attributes atts, List<PluginWrapper.Dependency> optionalDependencies) {
```
Comment (daniel-beck) [inline core/src/main/java/hudson/ClassicPluginStrategy.java]: Moving code around makes diffs difficult to read.

**Comment by daniel-beck** [comment]
From reading the code, this looks like :+1: to me. Now on to the testing…

File: core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java [hunk 17]
```diff
@@ -379,19 +378,30 @@ private User createAccount(StaplerRequest req, StaplerResponse rsp, boolean self
...
-            LOGGER.log(Level.WARNING, "Failed to set the e-mail address",e);
+        if(isMailerPluginPresent()) {
+            try {
+                // legacy hack. mail support has moved out to a separate plugin
+                Class<?> up = Jenkins.getInstance().pluginManager.uberClassLoader.loadClass("hudson.tasks.Mailer$UserProperty");
+                Constructor<?> c = up.getDeclaredConstructor(String.class);
+                user.addProperty((UserProperty)c.newInstance(si.email));
+            } catch (ReflectiveOperationException e) {
+                throw new RuntimeException(e);
+            }
```
Comment (kwhetstone) [inline core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java]: Maybe log a message here with the attempted action.

File: core/src/main/java/hudson/PluginManager.java (hunk 2 above)
Comment (kwhetstone) [inline core/src/main/java/hudson/PluginManager.java]: Hey, adding doc is always welcome.

File: core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java (hunk 17 above)
Comment (kzantow) [inline core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java]: It's not attempting anything, it doesn't proceed into this block unless isMailerPluginPresent() is true, in which case the class check is already done and instead of silently failing as it was doing, it will throw an exception.

**Comment by kwhetstone** [comment]
:bee:

**Comment by daniel-beck** [comment]
Seems pretty solid in testing. It's annoying as hell, but that's how implied dependencies work…
